### PR TITLE
Integrating MAYGEN project to structgen package

### PR DESCRIPTION
**Pull request by MehmetAzizYirik**
- Integrating maygen source and test files to structgen
- Improved also the SDFWriter class

### CODE REVIEW

**Review by egonw — COMMENTED** [review]
I left some initial comments.

File: storage/ctab/src/main/java/org/openscience/cdk/io/SDFWriter.java
Comment (egonw) [inline storage/ctab/src/main/java/org/openscience/cdk/io/SDFWriter.java:276]: @johnmay, you may want to check out the changes in this file.

Comment (johnmay) [reply storage/ctab/src/main/java/org/openscience/cdk/io/SDFWriter.java:276]: They look fine, I had a similar patching waiting to find time to apply

**Comment by johnmay** [comment]
All looks good to me, I think I may tweak some things like naming once merged (i.e. just keep it under ../cdk/structgen/ rather than add another subpackage). But that can all happen after being applied providing it's before release.

**Comment by egonw** [comment]
> All looks good to me, I think I may tweak some things like naming once merged (i.e. just keep it under ../cdk/structgen/ rather than add another subpackage). But that can all happen after being applied providing it's before release.

@johnmay, we wait for a bit more JavaDoc? or also after merge?

**Comment by johnmay** [comment]
Don't mind - I guess some would be good. Also just seen the MIT license, so we going to dual license Egon? Sounds complicated

**Comment by egonw** [comment]
> Don't mind - I guess some would be good. Also just seen the MIT license, so we going to dual license Egon? Sounds complicated

we already have other licenses in the CDK. I agree we don't communicate this well. something to think about. this is why copyright/license headers in the source files help.

**Comment by MehmetAzizYirik** [comment]
> > Don't mind - I guess some would be good. Also just seen the MIT license, so we going to dual license Egon? Sounds complicated
> 
> we already have other licenses in the CDK. I agree we don't communicate this well. something to think about. this is why copyright/license headers in the source files help.

If possible and if needed, we can also change the license as how you wish. I am currently writing the javadocs and will psuh changes to the pull request today.

**Comment by MehmetAzizYirik** [comment]
@johnmay @egonw I committed my changes. Added the javadoc and some other small changes as you requested. If I need to do anything else, please let me know. Thanks!

**Review by egonw — APPROVED** [review]

**Comment by egonw** [comment]
@MehmetAzizYirik, okay, getting compile errors: https://github.com/cdk/cdk/runs/4372594354?check_suite_focus=true

**Comment by johnmay** [comment]
Looks like Maygen.java is trying to import the Commons CLI package. Is the command line component still needed?

Not a blocker (as Egon says it's a mess already :-)) but I think it's cleaner to relicense as LGPL as it just makes things simpler to keep track of.

**Comment by MehmetAzizYirik** [comment]
> Looks like Maygen.java is trying to import the Commons CLI package. Is the command line component still needed? (build failed)
> 
> Not a blocker (as Egon says it's a mess already :-)) but I think it's cleaner to relicense as LGPL as it just makes things simpler to keep track of.

@egonw @johnmay Yes the compiling issue is solved, Valentyn ( @javadev ) updated the dependencies. I committed that changes.

**Comment by egonw** [comment]
> Looks like Maygen.java is trying to import the Commons CLI package. Is the command line component still needed? (build failed)

Personally, I would prefer CLI tools in a separate module. @johnmay, what are your ideas about that?

> Not a blocker (as Egon says it's a mess already :-)) 

Did I say that? :) No, I would not say "mess"... we just have a few source files that are MIT and/or public domain.

> but I think it's cleaner to relicense as LGPL as it just makes things simpler to keep track of.

Yes, agreed.

**Comment by johnmay** [comment]
Yes CLI separately - prompts me actually to just add some other CLIs that I have floating around, I’ll try and find time tonight to fix this up.

**Comment by egonw** [comment]
> prompts me actually to just add some other CLIs that I have floating around, I’ll try and find time tonight to fix this up.

I'd love that!

**Comment by johnmay** [comment]
Not got time tonight but hopefully tomorrow. The dependency changes are very suspect,  i.e. structgen depending on cdk-pdb and cdk-group. May be the case but seems odd. Likewise the Vecmath should be transient via the cdk-interfaces since it's mainly for the Point2d etc. As noted, we are a library so adding a dependencies on commons-cli sets this further back.

**Comment by johnmay** [comment]
Closing due - now #801

**Comment by MehmetAzizYirik** [comment]
> Okay @MehmetAzizYirik reopened this one as requested. I have rebased here and am working on it here: https://github.com/cdk/cdk/tree/maygen.
> 
> List may grow but here is things I will be doing:
> 
> * [x]  rebased on current master
> * [x]  avoid new Atom() should use building
> * [ ]  Pass in the builder, prefer silent builder
> * [ ]  remove commons-cli, may be able to put in a separate source root
> * [ ]  remove commons-lang3 usage
> * [ ]  remove ctab/smiles/sdg compile dependency, IO should be indepandant. Probably needs a "collector" like API
> * [ ]  stop tests writing files to disk
> * [ ]  faster tests, we have the "SlowTest" tag for things we don't want to run every build but now takes 9 seconds on my laptop. The entire test suite only takes 60 seconds, so not ideal :-). This probably just means simpler unit tests.
> * [ ]  Use atomic numbers rather than Strings (optional)

@johnmay Thanks John for the detailed list. Since you said " I ll be doing", I dont wanna do changes without informing. For now if that is ok for you, I can take a look at the last three in the list ?

**Comment by johnmay** [comment]
Because I needed to rebase due to the large number of changes (related to Log4j and dependency purging) I can't merge back into this branch so will create another PR - sound okay? I think just the test fixing would be good, the atomic number is a nice to have and probably quite involved.

**Comment by MehmetAzizYirik** [comment]
> Because I needed to rebase due to the large number of changes (related to Log4j and dependency purging) I can't merge back into this branch so will create another PR - sound okay? I think just the test fixing would be good, the atomic number is a nice to have and probably quite involved.

@johnmay Absolutely okay whatever you recommend. :) I will look at the test fixing.

**Comment by johnmay** [comment]
Unrelated to anything that needs changing but could you explain this logic. I do most multithreading in C/C++/pthreads now a days so a little rusty on ForkJoinPool but this looks like it creates a redundant thread.

```java
try {
    new ForkJoinPool(size)
            .submit(
                    () ->
                            newDegrees
                                    .parallelStream()
                                    .forEach(new Generation(this)::run))
            .get();
} catch (InterruptedException | ExecutionException ex) {
    if (verbose) {
        Logger.getLogger(Maygen.class.getName())
                .log(Level.SEVERE, ex, () -> "Formula " + localFormula);
    }
    Thread.currentThread().interrupt();
}
```

```java
try {
    new ForkJoinPool(size)
            .submit(
                    () ->
                            newDegrees
                                    .parallelStream()
                                    .forEach(new Generation(this)::run))
            .get();
} catch (InterruptedException | ExecutionException ex) {
    if (verbose) {
        Logger.getLogger(Maygen.class.getName())
                .log(Level.SEVERE, ex, () -> "Formula " + localFormula);
    }
    Thread.currentThread().interrupt();
}
```

i.e. I believe it's equivalent to this:
```java
                Thread t = new Thread(() ->
                                        newDegrees
                                                .parallelStream()
                                                .forEach(new Generation(this)::run)
                );
                t.start();
                t.join();
```
which other than spawning a redundant thread is the same as just calling:

```java
newDegrees .parallelStream().forEach(new Generation(this)::run)
```

**Comment by johnmay** [comment]
```
newDegrees .parallelStream().forEach(new Generation(this)::run)
```
runs faster in the tests :p

**Comment by johnmay** [comment]
But only slightly

**Comment by johnmay** [comment]
@MehmetAzizYirik Could you check this commit looks reasonable: https://github.com/cdk/cdk/pull/811/commits/a0dfd2b9ec44497850df6793d26f248c02947960. Providing the logic is correct I can now move the IO outside of Maygen

**Comment by MehmetAzizYirik** [comment]
> Unrelated to anything that needs changing but could you explain this logic. I do most multithreading in C/C++/pthreads now a days so a little rusty on ForkJoinPool but this looks like it creates a redundant thread.
> 
> ```java
> try {
>     new ForkJoinPool(size)
>             .submit(
>                     () ->
>                             newDegrees
>                                     .parallelStream()
>                                     .forEach(new Generation(this)::run))
>             .get();
> } catch (InterruptedException | ExecutionException ex) {
>     if (verbose) {
>         Logger.getLogger(Maygen.class.getName())
>                 .log(Level.SEVERE, ex, () -> "Formula " + localFormula);
>     }
>     Thread.currentThread().interrupt();
> }
> ```
> 
> i.e. I believe it's equivalent to this:
> 
> ```java
>                 Thread t = new Thread(() ->
>                                         newDegrees
>                                                 .parallelStream()
>                                                 .forEach(new Generation(this)::run)
>                 );
>                 t.start();
>                 t.join();
> ```
> 
> which other than spawning a redundant thread is the same as just calling:
> 
> ```java
> newDegrees .parallelStream().forEach(new Generation(this)::run)
> ```

We  just parallelized MAYGEN fork join after reading some sources but this can be even changed if there is a better option. The logic is, after getting the pre-hydrogen distribution, we have a new list of degrees and we parallelized MAYGEN based on this new degree list. But again, if you think that can be changed, please feel free to do that changes. I am open to learn better methods because I dont think this parallelization is efficient enough.

**Comment by MehmetAzizYirik** [comment]
> @MehmetAzizYirik Could you check this commit looks reasonable: [a0dfd2b](https://github.com/cdk/cdk/commit/a0dfd2b9ec44497850df6793d26f248c02947960). Providing the logic is correct I can now move the IO outside of Maygen

@johnmay Thanks a lot John! I went though the changes and looks cleaner and perfect!

**Comment by johnmay** [comment]
Awesome almost done then, just need to move the IO to a separate functor. The tests could simply be tagged as "slow" in which case they don't get run as part of the main suite. Ideally we want to run as many tests as possible for each build but it's okay if some are only run occasionally. 

Using an Executor service may be better for the multi-threading but I need to look more at the code and profile. For now removing the redundant ForkJoinPool saves some time on the tests at least.

Is it possible, to reuse a Maygen object? May be able to make the tests faster using simple parameterisation.

**Comment by johnmay** [comment]
In-fact SonarCloud even recommends using Parameterized tests :-).

**Comment by johnmay** [comment]
Infact the simplest thing to do is not run the tests twice :-)

```java
        Maygen maygen = new Maygen(SilentChemObjectBuilder.getInstance());
        maygen.setFormula("C3O6PH5");
        maygen.run();
        assertEquals(51323, maygen.getCount());
        maygen.setMultiThread(true);
        maygen.run();
        assertEquals(51323, maygen.getCount());
```

**Comment by MehmetAzizYirik** [comment]
> Infact the simplest thing to do is not run the tests twice :-)
> 
> ```java
>         Maygen maygen = new Maygen(SilentChemObjectBuilder.getInstance());
>         maygen.setFormula("C3O6PH5");
>         maygen.run();
>         assertEquals(51323, maygen.getCount());
>         maygen.setMultiThread(true);
>         maygen.run();
>         assertEquals(51323, maygen.getCount());
> ```

Well actually it is testing both the single and multithread within the same tests rather than making them separate. If you recommend something else for testi multithread, we can modify that is also fine.

**Comment by MehmetAzizYirik** [comment]
> Awesome almost done then, just need to move the IO to a separate functor. The tests could simply be tagged as "slow" in which case they don't get run as part of the main suite. Ideally we want to run as many tests as possible for each build but it's okay if some are only run occasionally.
> 
> Using an Executor service may be better for the multi-threading but I need to look more at the code and profile. For now removing the redundant ForkJoinPool saves some time on the tests at least.
> 
> Is it possible, to reuse a Maygen object? May be able to make the tests faster using simple parameterisation.

About parallelization, it needs better approach actually and we can change it as how you recommend. 

About the parameterisation, I agree it will be better if we reuse the maygen object.

**Comment by johnmay** [comment]
I've made the parameterized tests now and actually allows us to mark the single threaded as slow. And only takes 1s or so to run them all multithreaded

**Comment by johnmay** [comment]
Now #811

**Pull request closed without merging**
